model/invoice.rb

```diff
@@ -15,58 +15,69 @@ def name
   end
 
   def charge
+    reload # Reload to get the latest status to avoid double charging
     unless (Stripe.api_key = Config.stripe_secret_key)
-      puts "Billing is not enabled. Set STRIPE_SECRET_KEY to enable billing."
-      return
+      Clog.emit("Billing is not enabled. Set STRIPE_SECRET_KEY to enable billing.")
+      return true
     end
 
     if status != "unpaid"
-      puts "Invoice[#{ubid}] already charged: #{status}"
-      return
+      Clog.emit("Invoice already charged.") { {invoice_already_charged: {ubid: ubid, status: status}} }
+      return true
     end
 
-    if content["cost"] < Config.minimum_invoice_charge_threshold
+    amount = content["cost"].to_f.round(2)
+    if amount < Config.minimum_invoice_charge_threshold
       update(status: "below_minimum_threshold")
-      puts "Invoice[#{ubid}] cost is less than minimum charge cost: $#{content["cost"]}"
-      return
+      Clog.emit("Invoice cost is less than minimum charge cost.") { {invoice_below_threshold: {ubid: ubid, cost: amount}} }
+      return true
     end
 
-    unless (billing_info = BillingInfo[content.dig("billing_info", "id")])
-      puts "Invoice[#{ubid}] doesn't have billing info"
-      return
+    unless (billing_info = BillingInfo[content.dig("billing_info", "id")]) && billing_info.payment_methods.count > 0
```

Comment:
It is confusing to read unless statements with multiple conditions. How about;

```
if (billing_info = BillingInfo[content.dig("billing_info", "id")]).nil? || billing_info.payment_methods.count == 0
```

Reply:
I changed it